### PR TITLE
Hotfix/start build latency optimizations

### PR DESCRIPTION
**Pull request by mariangemarcano**
- Add support for global processing of builds by priority 
- Optimizations for better cpu resource usage
- Fixes to speed up starting builds
- Add support for cancel build requests on resume
- Order finished builds by complete time
- Display pending builds order by priority 
- Fixes / performance improvements when displaying pending builds in the builder’s page 
- Some fixes to be able to run tests under pypy

### CODE REVIEW

**Comment by campos-ddc** [comment]
Code looks fine to me, lots of tests sounds good too.

Can't tell for sure because it is such a large change, but ok for me

File: master/buildbot/db/buildrequests.py [hunk 1]
```diff
@@ -220,45 +313,107 @@ def getResults(query):
         return self.db.pool.do(thd)
 
     @with_master_objectid
-    def getOldestBuildRequestInQueue(self, buildername, _master_objectid=None):
+    def getPrioritizedBuildRequestsInQueue(self, queue, buildername=None, sourcestamps=None,
```
Comment (opdude) [inline master/buildbot/db/buildrequests.py]: Can you add some documentation to clear up what this is used for.

Also is there any time when you don't want the order? Maybe we don't need to offer the option?

File: master/buildbot/db/buildrequests.py [hunk 2]
```diff
@@ -397,31 +567,31 @@ def thd(conn):
...
-                q = sa.select([buildrequests_tbl.c.artifactbrid]) \
-                    .where(id == requests[0].id)
-                res = conn.execute(q)
-                row = res.fetchone()
-                # by default it will mark using artifact generated from merged brid
-                stmt2 = buildrequests_tbl.update() \
-                    .where(buildrequests_tbl.c.id.in_(mergedrequests)) \
-                    .values(artifactbrid=requests[0].id)\
-                    .values(mergebrid=requests[0].id)
+        # we'll need to batch the brids into groups of 100, so that the
```
Comment (opdude) [inline master/buildbot/db/buildrequests.py]: Next time try to keep format changes to a minimum or in another PR so it makes it easier to see that this is not part of the changes

File: master/buildbot/process/builder.py [hunk 3]
```diff
@@ -132,7 +136,8 @@ def getOldestRequestTime(self):
 
         @returns: datetime instance or None, via Deferred
         """
-        unclaimed = yield self.master.db.buildrequests.getOldestBuildRequestInQueue(buildername=self.name)
+        unclaimed = yield self.master.db.buildrequests.getBuildRequests(
+            buildername=self.name, claimed=False)
```
Comment (opdude) [inline master/buildbot/process/builder.py]: Could this fit on a single line or is it too long for PEP8?

File: master/buildbot/process/buildrequestdistributor.py [hunk 4]
```diff
@@ -351,9 +363,46 @@ def canStartBuild(self, slave, breq):
...
+    def setupNextBuildRequest(self, bldr, breq):
+        # by default katana merges buildrequests
+        if bldr.config.mergeRequests is None:
+            bldr.config.mergeRequests = True
+
+        self.bldr = bldr
+
+        self.nextSlave = self.bldr.config.nextSlave
+        if not self.nextSlave:
+            self.nextSlave = lambda _,slaves: random.choice(slaves) if slaves else None
```
Comment (opdude) [inline master/buildbot/process/buildrequestdistributor.py]: One day we should make this a bit more intelligent ;)

Comment (mariangemarcano) [inline master/buildbot/process/buildrequestdistributor.py]: sure thing!

File: master/buildbot/process/buildrequestdistributor.py [hunk 5]
```diff
@@ -364,6 +413,134 @@ def claimBuildRequests(self, breqs):
...
+        """
+        Finds the next builder that should start running builds
+        selecting the one with the higher priority and oldest submitted time.
+
+        it will return None if there are no slaves available for the task.
+
+        @param queue: None will select the higher priority overall pending buids,
+        if queue is 'unclaimed' will select only the pending builds and if queue='resume'
+        it will select only builds pending to be resume
+        @returns: a build request dictionary or None via Deferred
```
Comment (opdude) [inline master/buildbot/process/buildrequestdistributor.py]: Lovely :)!

File: master/buildbot/process/buildrequestdistributor.py [hunk 6]
```diff
@@ -751,9 +834,8 @@ class BuildRequestDistributor(service.Service):
     methods.
     """
 
-    # todo read from configuration if want to use buildbot's or katana build chooser
-    BuildChooser = KatanaBuildChooser
-    
+    BuildChooser = BasicBuildChooser
```
Comment (opdude) [inline master/buildbot/process/buildrequestdistributor.py]: Gone back to the basicbuildchooser?

Comment (mariangemarcano) [inline master/buildbot/process/buildrequestdistributor.py]: Yes that's buildbot's BuildRequestDistributor and we are not using it anymore

File: www/templates/build.html [hunk 7]
```diff
@@ -23,36 +23,45 @@
...
-            {{ forms.stop_build(build_url+"/stop"+codebases_arg, authz, on_all=False, short=False, label='This Build')
-            }}
-            <div class="divider"></div>
-            {% endif %}
+            {% if b.isResuming() %}
+                <div class="divider"></div>
+                <h2 class="head-2">
+                        Stop Build
+                </h2>
+                    {{ forms.stop_build(build_url+"/cancel"+codebases_arg, authz, on_all=False, short=False, label='This Build')
```
Comment (opdude) [inline www/templates/build.html]: Maybe we should create a jinja macro for this passing in the url difference and the label?

**Comment by opdude** [comment]
Looks sane overall, reviewed the backend changes and the frontend changes. Didn't review the tests in any great detail however.

**Pull request merged**